### ISSUE TITLE
boot time is slow because of lack of random entropy from host

### ISSUE THREAD

**Issue by ColinIanKing** [canonical/multipass#709]
Most VM hosts can pass entropy to the VM and hence allow the kernel to get entropy really fast during the early kernel boot.   For example kvm-qemu supports this feature.  However, we don't seem to have this on the Windows multipass startups, causing long delays in boot:

For example, on xenial VMs on multipass for Windows we get a 3.47 second delay between the input init and the random pool init (from Xenial dmesg):

[    7.280017] input: AT Translated Set 2 keyboard as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A03:00/device:07/VMBUS:01/d34b2567-b9b6-42b9-8778-0a4ec0b955bf/serio2/input/input5
[   10.756222] random: nonblocking pool is initialized

And the same for Bionic too, we get an 8.34 second delay:

[    7.058739] input: AT Translated Set 2 keyboard as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A03:00/device:07/VMBUS:01/d34b2567-b9b6-42b9-8778-0a4ec0b955bf/serio2/input/input5
[   15.396069] random: crng init done

There probably is a way to do with the Windows hypervisor and such a fix would radically reduce boot times.

**Comment by ricab** [comment]
Thank you for the suggestion, this is something worth looking into indeed.

**Comment by ColinIanKing** [comment]
Of course, the solution maybe to use the linux-image-azure kernel instead of the generic kernel.

**Comment by tobe2098** [comment]
Issue is blocked until the Hyper-V refactor PR, as it is related #4080 
Apparent options for this problem:
- Manage RNG hardware passthrough (like vTPM, but only in Shielded VMs)
- Installing rng-tools in the guest and running `rngd` from boot, maybe add a EGD-compatible server on the host. This solution seems very intrusive since we are adding a software package inside the guest, and adding a possible Windows dependency since CNG does not support egd-compatible servers. An alternative would be a Powershell script like:
```
Add-Type -AssemblyName System.Security

$listener = [System.Net.Sockets.TcpListener]::new([System.Net.IPAddress]::Any, 7111)
$listener.Start()
Write-Host "EGD server running on port 7111..."

while ($true) {
    $client = $listener.AcceptTcpClient()
    $stream = $client.GetStream()

    # Wait for 1-byte request
    $req = $stream.ReadByte()
    if ($req -gt 0) {
        $buf = New-Object byte[] ($req)
        [System.Security.Cryptography.RandomNumberGenerator]::Create().GetBytes($buf)
        $stream.Write($buf, 0, $buf.Length)
    }

    $client.Close()
}
```
Which would provide the necessary entropy to the `rngd`.
